src/core/configuration/DefaultConfig.ts

```diff
@@ -143,11 +150,36 @@ export abstract class DefaultServerConfig implements ServerConfig {
     return 60 * 1000;
   }
 
-  lobbyMaxPlayers(map: GameMapType, mode: GameMode): number {
+  protected maxPlayersForMap(map: GameMapType): number {
     const [l, m, s] = numPlayersConfig[map] ?? [50, 30, 20];
     const r = Math.random();
     const base = r < 0.3 ? l : r < 0.6 ? m : s;
-    return mode === GameMode.Team ? Math.ceil(base * 1.5) : base;
+    return base;
+  }
+
+  calcPlayerConfig(map: GameMapType, mode: GameMode): ServerPlayerConfig {
+    const maxPlayerForMap = this.maxPlayersForMap(map);
+
+    switch (mode) {
+      case GameMode.FFA:
+        return {
+          maxPlayers: maxPlayerForMap,
+          numPlayerTeams: undefined,
+        };
+
+      case GameMode.Team: {
+        // between 2 and 6 teams
+        const numPlayerTeams = 2 + Math.floor(Math.random() * 5);
+        const teamModeMaxPlayerCount = Math.ceil(maxPlayerForMap * 1.5);
+
+        const unevenPlayersPerTeam = teamModeMaxPlayerCount / numPlayerTeams;
+        const evenPlayersPerTeam = Math.floor(unevenPlayersPerTeam);
+
+        const maxPlayers = evenPlayersPerTeam * numPlayerTeams;
```

Comment:
This can be simplified to:
```suggestion
        const maxPlayers= teamModeMaxPlayerCount - (teamModeMaxPlayerCount % numPlayerTeams);
```

Reply:
Yes, it's fewer lines, but I don't think that's simpler, nor more maintainable. It's more cryptic. I imagine a poll of your average web developer will have them scratching their heads at that vs the obvious way.

Reply:
I want to push back a bit on the idea that `%` is cryptic or obscure. Modulo is not some niche mathematical concept. It's a fundamental arithmetic operation provided by the hardware, so it is widely used in computer science. In this case, it’s actually more efficient, because when you do `a / b`, the CPU calculates both the quotient and the remainder at the same time. In the original code, you throw away the known remainder, and re-calculate it through a different method. This is not just a stylistic choice - the suggestion uses fewer CPU instructions, and therefore runs faster.

This isn't a nitpick; it’s about using the tools we have efficiently. If we start avoiding common operators like `%` because they might confuse someone, we're setting a pretty low bar for what's considered readable. I’d argue that modulo is something every developer should be comfortable with, especially when it's exactly the right tool for the job.

Reply:
It's not specifically about modulo, it's about the fact that most people will have to look at it and figure out you're doing a division, taking the remainder, and then removing that reminder, all in one line. The previous way was blindingly obvious to anyone who understands arithmetic. That's because this is software engineers, not computer science. 

> In this case, it’s actually more efficient,

Firstly, you would need to prove that, as 2 operations vs 3 is hard to predict until run through a profiler. Given the aggressiveness of modern browser javascript interpreters I wouldn't be surprised if the compiler already spots the optimisation and makes it all a divmod anyway.

Secondly, this operation happens once every 60 seconds, so I don't care in the slightest about such a micro optimisation. I care about bugs and long term maintainability, not about feeling clever.

Reply:
Thanks for the thoughtful response, and I appreciate your focus on maintainability, this is definitely a priority we share.

That said, I want to zoom out a bit. This isn’t about feeling clever or shaving nanoseconds off something that runs once per minute. It’s about cultivating habits of writing code that’s both correct and efficient by design, even in places where performance doesn't yet matter.

You're right that modern JS engines are aggressive with optimization, and while it is possible that it may collapse `Math.floor(x / y) * y` into a `x - (x % y)` under the hood, it probably will not because, as you mentioned, this code does not run frequently enough for JIT optimizations to pay off. Regardless, we should never rely on a jitter to clean up inefficient algorithms that we could express more directly ourselves, especially when that direct form is semantically correct, idiomatic, and just as readable once you're comfortable with it.

And to directly address the concern: this is not complicated code. `x - (x % y)` is a clear and concise way to round down to the nearest multiple of y. The deeper point here is that `%` is not obscure, it's fundamental. If a developer doesn’t immediately understand what `x - (x % y)` does, that’s a growth opportunity, not a reason to avoid it.

I’d also argue that “blindingly obvious” is contextual. For someone familiar with modulo, `x - (x % y)` instantly communicates “round x down to nearest multiple of y,” which is exactly what we're trying to do. It’s not clever; it’s the approach we use throughout the code corpus in every location except this one.

So while the performance benefit here is minor, the principle is important: use the right operation for the job, trust the reader to understand common arithmetic, and remain consistent with how problems are solved throughout the codebase.